src/librustc_ast/token.rs

```diff
@@ -796,6 +742,31 @@ impl Nonterminal {
         }
         false
     }
+
+    // See issue #74616 for details
+    pub fn ident_name_compatibility_hack(
+        &self,
+        orig_span: Span,
+        source_map: &SourceMap,
+    ) -> Option<(Ident, bool)> {
+        if let NtIdent(ident, is_raw) = self {
+            let filename = source_map.span_to_filename(orig_span);
+            if let FileName::Real(RealFileName::Named(path)) = filename {
```

Comment:
Nit: the cheaper condition should go first and I suspect that `orig_span.ctxt().outer_expn_data()` is cheaper than the search through the source map.